File: Archived2/Java/Gold/9663.java
import java.io.*;

public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static boolean[][] board; // true: queen이 위치한다.
    private static int successCount = 0;
    private static int boardSize = 0;

    public static void main(String[] args) throws IOException {

        boardSize = Integer.parseInt(br.readLine());
        board = new boolean[boardSize][boardSize];

        for (int i=0;i<boardSize;i++) {
            board[0][i] = true;
            setQueens(0,i);
            board[0][i] = false;
        }

        System.out.println(successCount);
        br.close();
    }

    private static void setQueens(int y, int x) {

        if (isQueenInHorizontal(y, x)
                || isQueenInVertical(y, x)
                || isQueenInLeftDiagonal(y, x)
                || isQueenInRightDiagonal(y, x)) return;

        if (y == boardSize-1) {
            successCount += 1;
            return;
        }

        for (int i=0; i<boardSize; i++) {
            board[y+1][i] = true;
            setQueens(y+1, i);
            board[y+1][i] = false;
        }
    }

    private static boolean isQueenInHorizontal(int y, int x) {
        for (int i=0; i<boardSize; i++) {
            if (i == x) continue;
            if (board[y][i]) return true;
        }
        return false;
    }
    private static boolean isQueenInVertical(int y, int x) {
        for (int i=0; i<boardSize; i++) {
            if (i == y) continue;
            if (board[i][x]) return true;
        }
        return false;
    }

    private static boolean isQueenInLeftDiagonal(int y, int x) {
        // lower
        for (int i=1;i<boardSize; i++) {
            if (y+i >= boardSize || x-i < 0) break;
            if (board[y+i][x-i]) return true;
        }

        // upper
        for (int i=1; i<boardSize; i++) {
            if (y-i < 0 || x+i >= boardSize) break;
            if (board[y-i][x+i]) return true;
        }

        return false;
    }

    private static boolean isQueenInRightDiagonal(int y, int x) {
        // lower
        for (int i=1; i<boardSize; i++) {
            if (y+i >= boardSize || x+i >= boardSize) break;
            if (board[y+i][x+i]) return true;
        }

        //upper
        for (int i=1; i<boardSize; i++) {
            if (y-i < 0 || x-i < 0) break;
            if (board[y-i][x-i]) return true;
        }
        return false;
    }

}

File: Archived2/Kotlin/Gold/9663.kt
import java.io.BufferedReader
import java.io.InputStreamReader

val br = BufferedReader(InputStreamReader(System.`in`))
lateinit var board: ArrayList<BooleanArray>
var boardSize = 0
var successCount = 0


fun main() {

    boardSize = br.readLine().toInt()
    board = ArrayList()
    repeat(boardSize) {
        board.add(BooleanArray(boardSize))
    }


    repeat(boardSize) {
        board[0][it] = true
        setQueen(0,it)
        board[0][it] = false
    }

    println(successCount)
}

fun setQueen(y: Int, x: Int) {

    if(
        isQueenInHorizontal(y,x)
        || isQueenInVertical(y,x)
        || isQueenInLeftDiagonal(y,x)
        || isQueenInRightDiagonal(y,x)) return

    if (y == boardSize-1) {
        successCount++
        return
    }

    repeat(boardSize) {
        board[y+1][it] = true
        setQueen(y+1, it)
        board[y+1][it] = false
    }

}

fun isQueenInHorizontal(y: Int, x: Int):Boolean {

    for (i in 0 until boardSize) {
        if (i == x) continue
        if (board[y][i]) return true
    }
    return false

}

fun isQueenInVertical(y: Int, x: Int): Boolean {
    for (i in 0 until boardSize) {
        if (i == y) continue
        if (board[i][x]) return true
    }
    return false
}

fun isQueenInLeftDiagonal(y: Int, x: Int): Boolean {
    // lower
    for (i in 1 until boardSize) {
        if (y+i >= boardSize || x-i < 0) break
        if (board[y+i][x-i]) return true
    }

    // upper
    for (i in 1..boardSize) {
        if (y-i < 0 || x+i >= boardSize) break
        if (board[y-i][x+i]) return true
    }

    return false
}

fun isQueenInRightDiagonal(y: Int, x: Int): Boolean {
    // lower
    for (i in 1..boardSize) {
        if (y+i >= boardSize || x+i >= boardSize) break
        if (board[y+i][x+i]) return true
    }

    //upper
    for (i in 1 until boardSize) {
        if (y-i < 0 || x-i < 0) break
        if (board[y-i][x-i]) return true
    }
    return false
}